File: Controllers/AccessController.cs
using AccessManagementWebApp.Models;
using AccessManagementWebApp.Repositories;
using Microsoft.AspNetCore.Mvc;

namespace AccessManagementWebApp.Controllers
{
    public class AccessController : Controller
    {
        private readonly IUserAccessRepository _repository;

        public AccessController(IUserAccessRepository repository)
        {
            _repository = repository;
        }

        public IActionResult Index()
        {
            return View();
        }

        [HttpGet("createAccessForm.aspx")]
        [HttpGet("Access/CreateAccessForm")]
        public IActionResult CreateAccessForm()
        {
            return View();
        }

        [HttpPost]
        public async Task<IActionResult> RequestAccess(CreateAccessRequestModel model)
        {
            if (!ModelState.IsValid)
            {
                return View("CreateAccessForm", model);
            }

            var existingAccess = await _repository.CheckUserAccessAsync(model.Username, model.FunctionName);

            if (existingAccess != null)
            {
                ModelState.AddModelError(string.Empty, "User already has access to this function.");
                return View("CreateAccessForm", model);
            }

            var created = await _repository.CreateAccessAsync(model.Username, model.FunctionName);
            if (!created)
            {
                ModelState.AddModelError(string.Empty, "Unable to create access request.");
                return View("CreateAccessForm", model);
            }

            ViewBag.SuccessMessage = "Access request created successfully.";
            return View("CreateAccessForm");
        }

        [HttpGet("Access/RevokeAccessForm")]
        [HttpGet("revokeAccessForm.aspx")]
        public IActionResult RevokeAccessForm()
        {
            return View();
        }

        [HttpPost]
        public async Task<IActionResult> RevokeAccess(RevokeAccessRequestModel model)
        {
            if (!ModelState.IsValid)
            {
                return View("RevokeAccessForm", model);
            }

            var existingAccess = await _repository.CheckUserAccessAsync(model.Username, model.FunctionName);
            if (existingAccess == null)
            {
                ModelState.AddModelError(string.Empty, "User does not have access to this function.");
                return View("RevokeAccessForm", model);
            }

            var revoked = await _repository.RevokeAccessAsync(model.Username, model.FunctionName);
            if (!revoked)
            {
                ModelState.AddModelError(string.Empty, "Unable to revoke access.");
                return View("RevokeAccessForm", model);
            }

            ViewBag.SuccessMessage = "Access revoked successfully.";
            return View("RevokeAccessForm");
        }

        [HttpGet("Access/ChangePermissionForm")]
        [HttpGet("changePermissionForm.aspx")]
        public IActionResult ChangePermissionForm()
        {
            return View();
        }

        [HttpPost]
        public async Task<IActionResult> ChangePermission(ChangePermissionRequestModel model)
        {
            if (!ModelState.IsValid)
            {
                return View("ChangePermissionForm", model);
            }

            var existingAccess = await _repository.CheckUserAccessAsync(model.Username, model.FunctionName);
            if (existingAccess == null)
            {
                ModelState.AddModelError(string.Empty, "User does not have access to this function.");
                return View("ChangePermissionForm", model);
            }

            var updated = await _repository.UpdateAccessAsync(model.Username, model.FunctionName, model.NewPermission);
            if (!updated)
            {
                ModelState.AddModelError(string.Empty, "Unable to update permission.");
                return View("ChangePermissionForm", model);
            }

            ViewBag.SuccessMessage = "Access permissions updated successfully.";
            return View("ChangePermissionForm");
        }

        [HttpGet("reactivateAccessForm.aspx")]
        [HttpGet("Access/ReactivateAccessForm")]
        public IActionResult ReactivateAccessForm()
        {
            return View();
        }

        [HttpPost]
        public async Task<IActionResult> ReactivateAccess(ReactivateAccessRequestModel model)
        {
            if (!ModelState.IsValid)
            {
                return View("ReactivateAccessForm", model);
            }

            var existingAccess = await _repository.CheckUserAccessAsync(model.Username, model.FunctionName);
            if (existingAccess != null)
            {
                ModelState.AddModelError(string.Empty, "User already has active access to this function.");
                return View("ReactivateAccessForm", model);
            }

            var reactivated = await _repository.ReactivateAccessAsync(model.Username, model.FunctionName);
            if (!reactivated)
            {
                ModelState.AddModelError(string.Empty, "Unable to reactivate access.");
                return View("ReactivateAccessForm", model);
            }

            ViewBag.SuccessMessage = "Access reactivated successfully.";
            return View("ReactivateAccessForm");
        }
    }
}

File: src/controllers/accessController.js
class AccessController {
    async createAccessRequest(req, res) {
        // Logic to create a new access request
        const { userId, applicationId } = req.body;
        // Check if user already has access
        const userModel = new UserModel();
        const hasAccess = await userModel.checkUserAccess(userId, applicationId);
        
        if (hasAccess) {
            return res.status(400).json({ message: "User already has access to this application." });
        }

        // Logic to create access
        // Assume createAccess is a method in UserModel
        await userModel.createAccess(userId, applicationId);
        return res.status(201).json({ message: "Access request created successfully." });
    }

    async revokeAccessRequest(req, res) {
        // Logic to revoke access
        const { userId, applicationId } = req.body;
        const userModel = new UserModel();
        const hasAccess = await userModel.checkUserAccess(userId, applicationId);
        
        if (!hasAccess) {
            return res.status(400).json({ message: "User does not have access to this application." });
        }

        // Logic to revoke access
        await userModel.revokeAccess(userId, applicationId);
        return res.status(200).json({ message: "Access revoked successfully." });
    }

    async changeAccessPermission(req, res) {
        // Logic to change access permissions
        const { userId, applicationId, newPermissions } = req.body;
        const userModel = new UserModel();
        const hasAccess = await userModel.checkUserAccess(userId, applicationId);
        
        if (!hasAccess) {
            return res.status(400).json({ message: "User does not have access to this application." });
        }

        // Logic to update permissions
        await userModel.updateUserAccess(userId, applicationId, newPermissions);
        return res.status(200).json({ message: "Access permissions updated successfully." });
    }
}

module.exports = AccessController;